design/design-doc-custom-composition.md

```diff
@@ -0,0 +1,179 @@
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: CustomComposition
+metadata:
+  name: acmenetwork.acme.org
+spec:
+  writeConnectionSecretsToNamespace: crossplane-system
+  reclaimPolicy: Delete
+  compositeTypeRef:
+    apiVersion: aws.example.org/v1alpha1
+    kind: PrivateNetwork
+  image: acme/network-cdk8s:1.0.0
+```
+
+Once this `CustomComposition` is created, a new `Deployment` will be created by
+Crossplane with the following properties:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: acmenetwork.acme.org
+  labels:
+    app: acmenetwork.acme.org
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: acmenetwork.acme.org
+  template:
+    metadata:
+      labels:
+        app: acmenetwork.acme.org
+    spec:
+      containers:
+      - name: custom-composition
+        image: crossplane/custom-composition:0.0.1
+        env:
+        # COMPOSITE_TYPE will let the custom-composition controller know what type
+        # it should watch.
+        - name: COMPOSITE_TYPE
+          value: privatenetworks.aws.example.org
+        # CUSTOMCOMPOSITION_NAME is the name of the CustomComposition that will
+        # be sent over to composition engine alongside composite instance. 
+        - name: CUSTOMCOMPOSITION_NAME
+          value: acmenetwork.acme.org
+      - name: composition-engine
+        image: acme/network-cdk8s:1.0.0
+        ports:
+        - containerPort: 80
+```
+
+After this `Deployment` is up, the container named `custom-composition` will start
+watching `PrivateNetwork` types and it will reconcile only the ones with the following
+reference:
+```yaml
+apiVersion: acme.example.org
+kind: PrivateNetwork
+metadata:
+  name: something
+spec:
+  region: us-east-1
+  providerConfigRef:
+    name: my-creds
+  compositionRef:
+    apiVersion: apiextensions.crossplane.io/v1alpha1
+    kind: CustomComposition
+    name: acmenetwork.acme.org
+```
+
+In each of its reconciliation pass:
+1. It will make an HTTP request to `cdk-app` container
+with the `PrivateNetwork` instance and `CustomComposition` instance.
+  * The reason we include `CustomComposition` is that the cdk8s app might be configured
+    with some additional fields in the future or metadata of `CustomComposition`.
+2. The response from the cdk-app container is expected to be a set of valid
+    Kubernetes YAMLs.
+3. `custom-composition` will add owner references to all objects and run the
+   equivalent of `kubectl apply --prune -f <response YAMLs>`.
+
+At this point, we got all of our output deployed to the cluster.
+
+## Alternatives Considered
+
+### Separate Type for Each Custom Composition Type
+
+We could have `CDKComposition`, `HelmComposition` types with their own additional
+fields. But since for each instance of those types, a new `Deployment` is created,
+the entity that creates this `Deployment` couldn't be Crossplane because it'd not
+know of those types. So, for each type, users would have to install a CRD and a
```

Comment:
```suggestion
fields. For each instance of those types, a new `Deployment` is created, but
the entity that creates this `Deployment` couldn't be Crossplane because it would not
know about those types. So, for each type, users would have to install a CRD and a
```